includes/js/komoju-checkout-blocks.js

```diff
@@ -0,0 +1,99 @@
+const { useEffect, useCallback, useRef, createElement } = window.wp.element;
+
+function registerPaymentMethod(paymentMethod) {
+    let name = `${paymentMethod.id}`
+    const settings = window.wc.wcSettings.getSetting(`${name}_data`, {});
+
+    const komojuFields = createElement('komoju-fields', {
+        'token': '',
+        'name': 'komoju_payment_token',
+        'komoju-api': settings.komojuApi,
+        'publishable-key': settings.publishableKey,
+        'session': settings.session,
+        'payment-type': settings.paymentType,
+        'locale': settings.locale,
+        style: { display: 'none' },
+    });
+
+    const label = createElement('div', {
+        style: { display: 'block', alignItems: 'center', justifyContent: 'center', width: '100%' }
+    },
+        window.wp.htmlEntities.decodeEntities(settings.title || window.wp.i18n.__('NULL GATEWAY', 'test_komoju_gateway')),
+        createElement('img', {
+            src: settings.icon,
+            alt: settings.title || 'Payment Method Icon',
+            style: { display: 'flex', alignItems: 'center', justifyContent: 'center', marginLeft: '10px' }
+        }),
+        komojuFields
+    );
+
+    const KomojuComponent = ({ activePaymentMethod, emitResponse, eventRegistration }) => {
+        const { onPaymentSetup } = eventRegistration;
+        const komojuFieldEnabledMethods = ['komoju_credit_card', 'komoju_konbini', 'komoju_bank_transfer']
+
+        useEffect(() => {
+            const komojuField = document.querySelector(`komoju-fields[payment-type='${paymentMethod.paymentType}']`);
+            komojuField.style.display = 'block';
+
+            const unsubscribe = onPaymentSetup(async () => {
+                console.log('onPaymentSetup', paymentMethod.id, activePaymentMethod)
+                if (paymentMethod.id != activePaymentMethod) return;
+                if (!komojuFieldEnabledMethods.includes(paymentMethod.id)) return;
+
+                if (komojuField && typeof komojuField.submit === 'function') {
+                    var submitResult = await komojuField.broker.send({ type: 'submit' });
```

Comment:
Is there a reason we need to go directly through the broker like this?

Reply:
Yes, it is for below reasons. I tired use submit first but there were several problems. So I choose to use broker directly.
1. Place Order button is type of 'button' not 'submit'. Which means that checkout block does not calling the parent form.

```js
     // If this is part of a form submit, we want to add the token to the form and resubmit.
    if (result.token && event && this.formSubmitHandler) {
      const form = this.formSubmitHandler.form;

      // Now we add an input to the form with the token and submit it.
      const inputName = this.getAttribute('name') ?? 'komoju_token';
      let input = document.querySelector(`input[name="${inputName}"]`) as HTMLInputElement | null;
      if (!input) {
        input = document.createElement('input');
        input.type = 'hidden';
        input.name = inputName;
        form.append(input);
      }
      input.value = result.token.id;

      // Re-submit the form without our submit event getting in the way.
      this.submitParentForm();
      return;
    }
```

Previously in submit in komoju-fields.js, it make a new input form and attached to the form and send this data to the server. However, since they introduced new checkout block, they don't use `submit` but using `button` and handling event the wooCommerce way. This is why I am directly getting `submitResult` from broker and emitting event in the code below.

```js
                    const komoju_payment_token = submitResult.token.id;

                    return {
                        type: emitResponse.responseTypes.SUCCESS,
                        meta: {
                            paymentMethodData: {
                                komoju_payment_token
                            },
                        },
                    };
                }
```

2. To throw error
This could be improved if I understand how the event is dispatched and handled in submit 🤔 Below is the code form `komoju-field-element.ts`

```js
    // On error, emit a komoju-invalid event in case consumers want to handle it.
    if (result.errors) {
      this.dispatchEvent(new CustomEvent('komoju-invalid', {
        detail: { errors: result.errors }, bubbles: true, composed: true,
      }));
      return;
    }
```
All I needed was that I need to throw error directly after broker's send.

```js
var submitResult = await komojuField.broker.send({ type: 'submit' });

if (submitResult.errors && submitResult.errors.length > 0) {
    return {
        type: emitResponse.responseTypes.ERROR,
        message: submitResult.errors[0].message,
    };
}
```
If I don't `emitResponse` here, the checkout block's payment process will be proceed despite when it has error.

So, I tried to mimic the action how komoju-fields were doing and make it aligned with checkout blocks, is my idea.

Reply:
The code snippet you mentioned that creates an input element and such will not get called if you call `komojuField.submit()` directly (specifically because `event` is not present). You can just call `submit` and it will return the token object on success.

Very good point regarding errors though. I'll have to keep that in mind for improving the library :grin: 

Here's a slightly janky way to achieve what you want without hitting the broker directly.

```javascript
function submitFields(fields) {
    return new Promise(async (resolve, reject) => {
        fields.addEventListener("komoju-invalid", reject);
        const token = await fields.submit();
        fields.removeEventListener("komoju-invalid", reject);
        if (token) resolve(token);
    });
}

// then in your submit handler:

try {
    const token = await submitFields(komojuField);
    return {
        type: emitResponse.responseTypes.SUCCESS,
        meta: {
            paymentMethodData: {
                komoju_payment_token: token.id
            },
        },
    };
} catch(e) {
    return {
        type: emitResponse.responseTypes.ERROR,
        message: e.detail.errors[0].message,
    };
}
```

Anyway your current solution is not bad at all. It's just that the broker isn't part of the "public API", and so it'd be nice if we could reserve the freedom to change how it works under the hood without breaking this plugin.

Reply:
The code above look really clean! Thank you for the feedback.
Right, I get token by submitting it. My lack of understanding led me to write a hacky code but this seems definitely better 😸 

About broker, I was also having the same idea that referencing broker isn't a good idea. 

Thank you again for the feedback and I've updated & pushed the code with your suggestion.